src/CoreWCF.BuildTools/src/OperationInvokerGenerator.Emitter.cs

```diff
@@ -0,0 +1,233 @@
+                    {
+                        _builder.AppendLine($"{indentor}(({operationContractSpec.Method.ContainingType.ToDisplayString()})instance).{operationContractSpec.Method.Name}({string.Join(", ", invocationParams)});");
+                    }
+                    else
+                    {
+                        _builder.AppendLine($"{indentor}var result = (({operationContractSpec.Method.ContainingType.ToDisplayString()})instance).{operationContractSpec.Method.Name}({string.Join(", ", invocationParams)});");
+                    }
+                }
+
+                _builder.AppendLine($"{indentor}var outputs = AllocateOutputs();");
+
+                foreach (var (ouputIndex, parameterIndex, parameter) in outputParams)
+                {
+                    _builder.AppendLine($"{indentor}outputs[{ouputIndex}] = p{parameterIndex};");
+                }
+
+                if (isAsync)
+                {
+                    if (isTaskReturnType)
+                    {
+                        _builder.AppendLine($"{indentor}return (null, outputs);");
+                    }
+                    else
+                    {
+                        _builder.AppendLine($"{indentor}return (result, outputs);");
+                    }
+                }
+                else
+                {
+                    if (operationContractSpec.Method.ReturnsVoid)
+                    {
+                        _builder.AppendLine($"{indentor}return new ValueTask<(object, object[])>((null, outputs));");
+                    }
+                    else
+                    {
+                        _builder.AppendLine($"{indentor}return new ValueTask<(object, object[])>((result, outputs));");
+                    }
+                }
+
+                indentor.Decrement();
+                _builder.AppendLine($"{indentor}}}");
+                _builder.AppendLine();
+                _builder.Append($"{indentor}public object[] AllocateInputs() => ");
+                if (inputParameterCount == 0)
+                {
+                    _builder.AppendLine("Array.Empty<object>();");
+                }
+                else
+                {
+                    _builder.AppendLine($"new object[{inputParameterCount}];");
+                }
+                _builder.AppendLine();
+                _builder.Append($"{indentor}private object[] AllocateOutputs() => ");
+                if (outputParameterCount == 0)
+                {
+                    _builder.AppendLine("Array.Empty<object>();");
+                }
+                else
+                {
+                    _builder.AppendLine($"new object[{outputParameterCount}];");
+                }
+                _builder.AppendLine();
+
+                _builder.AppendLine($"{indentor}[System.Runtime.CompilerServices.ModuleInitializer]");
+                _builder.Append($"{indentor}internal static void RegisterOperationInvoker() => ");
```

Comment:
I'm concerned about having a separate static method for each and every registered operation. I've seen contracts with thousands of operations per interface. We should at least only have one module initializer per contract, but ideally just a single one for the entire project. It's all generated into a single source file so the `file` accessor means all of the generated types can be referenced from a single method.

Reply:
I kept a `RegisterOperationInvoker` static method per `OperationInvoker` implementation but this method is no longer attributed with `ModuleInitializer`. 
Instead An `OperationInvokerModuleInitializer` class is generated at the end of the file and this class exposes a unique `ModuleInitializer` attributed method that call all the `RegisterOperationInvoker` method of the generated `OperationInvoker` types.